orion/evaluation/contextual.py

```diff
@@ -0,0 +1,215 @@
+from orion.evaluation.common import (
+    _accuracy, _any_overlap, _f1_score, _overlap, _precision, _recall, _weighted_segment)
+
+
+def _overlap_segment(expected, observed, start=None, end=None):
+    tp, fp, fn = 0, 0, 0
+
+    observed_copy = observed.copy()
+
+    for expected_seq in expected:
+        found = False
+        for observed_seq in observed:
+            if _overlap(expected_seq, observed_seq):
+                if not found:
+                    tp += 1
+                    found = True
+                if observed_seq in observed_copy:
+                    observed_copy.remove(observed_seq)
+
+        if not found:
+            fn += 1
+
+    fp += len(observed_copy)
+
+    return None, fp, fn, tp
+
+
+def _contextual_partition(expected, observed, start=None, end=None):
+    edges = set()
+
+    if start is not None:
+        edges.add(start)
+
+    if end is not None:
+        edges.add(end)
+
+    for edge in expected + observed:
+        edges.update(edge)
+
+    partitions = list()
+    edges = sorted(edges)
+    last = edges[0]
+    for edge in edges[1:]:
+        partitions.append((last, edge))
+        last = edge
+
+    expected_parts = list()
+    observed_parts = list()
+    weights = list()
+    for part in partitions:
+        weights.append(part[1] - part[0])
+        expected_parts.append(_any_overlap(part, expected))
+        observed_parts.append(_any_overlap(part, observed))
+
+    return expected_parts, observed_parts, weights
+
+
+def _pad(lst):
+    return [(part[0], part[1] + 1) for part in lst]
+
+
+def contextual_confusion_matrix(expected, observed, data=None, start=None, end=None, method=None):
+    """Compute the confusion matrix between the ground truth and the detected anomalies.
+
+    Args:
+        expected (DataFrame or list of tuples):
+            Ground truth passed as a ``pandas.DataFrame`` or list containing
+            two columns: start and stop.
+        observed (DataFrame or list of tuples):
+            Detected anomalies passed as a ``pandas.DataFrame`` or list containing
+            two columns: start and stop.
+        data (DataFrame):
+            Original data, passed as a ``pandas.DataFrame`` containing timestamp.
+            Used to extract start and end.
+        start (int):
+            Minimum timestamp of the original data.
+        end (int):
+            Maximum timestamp of the original data.
+        method (function):
+            Function for computing confusion matrix (weighted segment, overlap segment).
+
+    Returns:
+        tuple:
+            number of true negative, false positive, false negative, true positive.
+    """
+
+    def ws(x, y, z, w):
+        return _weighted_segment(x, y, _contextual_partition, z, w)
+
+    if method is not None:
+        function = method
+    else:
+        function = ws
```

Comment:
Is there a better approach to define this?

I basically want `_weighted_segment` to be the default and `_overlap_segment` to be chosen when the user specifies they want the other approach.

Reply:
Instead of using a `method` argument, I would just have a boolean argument called `weighted=True`, documented as "If True (default), consider overlap weights when computing metrics. Otherwise, use simple boolean overlapping indicators." or similar.